### DISCUSSION TITLE
How to batch router.push calls?

### DISCUSSION THREAD

**Discussion by borispoehland** [vercel/next.js#44402] [Help]
Hello!

I am referring to this comment, where @ijjk suggested to use the discussions for this type of problem: https://github.com/vercel/next.js/issues/29240#issuecomment-923428999

So, in my application, I made a new `useQueryState` hook, where I update the URL whenever a state changes. Simplified, it looks like this:
```
import { useSearchParams, usePathname, useRouter } from "next/navigation";
import { useState } from "react";

export function useQueryState<S = string>(
  key: string,
  defaultValue?: S
) {
  const params = useSearchParams();
  const router = useRouter();
  const pathname = usePathname();

  const [value, setState] = useState<string>(
    params.get(key) ?? (defaultValue ? `${defaultValue}` : "")
  );

  function setValue(newValue: S) {
      router.replace(
        `${pathname}?${new URLSearchParams({
          ...Object.fromEntries(params),
          ...{
             [key]: newValue.toString(),
          }
        }).toString()}`
      );
    setState(newValue.toString());
  }

  return [
    value,
    setValue,
  ];
}
```

Then, in my app, I use them like this:
```
const [page, setPage] = useQueryState("page", 1);
const [sortBy, setSortBy] = useQueryState("sortBy", "name");

// later in a event handler
setPage(2)
setSortBy("price")
```

In this case, the query will not be updated for the page, since the `sortBy` query update interrupts it. In this specific case though, I don't want to update them together, because the way I designed this, every query param is independent and unknowing of other query params.

What I need is a form of batching, where I want the router to recognize that I want to update both query params, and updates them together.

Any advice is appreciated!

**Comment by icyJoseph** [comment]
Interesting. At first I thought that you were using a rather old React version because React now automatically batches state updates. However the issue is that you've got your own custom state modification. 

I think in this case there's a bit of overlap of state management, and at the same time, because next navigation now uses React internals, there seems to be a concurrent cancellation of the new planed state (page), in favour of sortBy. 

I'd say first that it might be better to use reducers for this but oh well, I felt like hacking something today. With ReactStrictMode off for better visibility:

```tsx
import { useSearchParams, usePathname, useRouter } from "next/navigation";

const foldIntoRecord = <T extends [string, string][]>(list: T) =>
  list.reduce((acc, [key, value]) => {
    acc[key] = value;
    return acc;
  }, {} as Record<string, string>);

const createQueryState = () => {
  const batch: [string, string][] = [];

  function useQueryState<State = string>(key: string, defaultValue?: State) {
    const params = useSearchParams();
    const router = useRouter();
    const pathname = usePathname();

    const value = params.get(key) ?? (defaultValue ? `${defaultValue}` : "");

    function setValue(newValue: State) {
      batch.push([key, `${newValue}`]);

      const current = foldIntoRecord(Array.from(params.entries()));

      const newQuery = {
        ...current,
        ...foldIntoRecord(batch),
      };

      queueMicrotask(() => {
        router.replace(
          `${pathname}?${new URLSearchParams(newQuery).toString()}`
        );

        batch.length = 0;
      });
    }

    return [value, setValue] as const;
  }

  return useQueryState;
};

export const useQueryState = createQueryState();
```

This should do the trick. 

Notice that there's zero dependency to React hooks. Rather we consume everything from the next navigation hooks, and we enqueue state updates. 

> The microtask is a short function which will run after the current task has completed its work and when there is no other code waiting to be run before control of the execution context is returned to the browser's event loop.
> 
> This lets your code run without interfering with any other, potentially higher priority, code that is pending, but before the browser regains control over the execution context, potentially depending on work you need to complete.
> 
> The importance of microtasks comes in its ability to perform tasks asynchronously but in a specific order. 


More info available at https://developer.mozilla.org/en-US/docs/Web/API/queueMicrotask.

So the two successive calls to router replace are queued, and we keep track of the batching. Once the sync code is done, these micro tasks run in order of batching, one after the other, so you never really see a React frame where the hooks are out of sync. 


https://user-images.githubusercontent.com/21013447/209863983-71b11bb1-1043-41fd-8b74-38b9eb562e5a.mov

Not sure this is the React-ish, or Next-js way of doing this, but why not? 😅

**Comment by franky47** [comment]
My library [`next-usequerystate`](https://github.com/47ng/next-usequerystate) has both batching and throttling features to avoid hitting rate limits of the Web History API, plus other kinds of goodies to sync the query string to React state.

Version 1.8.0 which brings support for the app router is in beta, feedback is welcome! 🙏
